public/js/controllers/blocks.js

```diff
@@ -0,0 +1,314 @@
+const conversionRate = 100000000;
+let blocks, mempoolSubsidy, blocksHolder;
+
+function calculateMaximumNumberOfBlocksToDisplay(blockElement) {
+    const blocksSection = $('.blocks-section');
+    const blocksSectionFirstChildHeight = blocksSection.children(":first").outerHeight(true);
+    const blocksSectionLastChildHeight = blocksSection.children(":last").outerHeight(true);
+    let blocksSectionHeight = blocksSection.outerHeight();
+
+    // block section should be at least same height as netstats section
+    const netstatsSectionHeight = $('.netstats-section').outerHeight();
+    if (netstatsSectionHeight > blocksSectionHeight) {
+        blocksSectionHeight = netstatsSectionHeight;
+    }
+
+    const totalAvailableWidth = blocksSection.width();
+    let totalAvailableHeight = blocksSectionHeight - blocksSectionFirstChildHeight - blocksSectionLastChildHeight;
+
+    // make blocks section fill available window height
+    const extraSpace = $(window).height() - $('#mainContainer').outerHeight(true);
+    if (extraSpace > 0) {
+        totalAvailableHeight += extraSpace;
+    }
+
+    const blockWidth = blockElement.width();
+    const blockHeight = blockElement.height() + 20; // for spacing between rows
+
+    const maxBlocksPerRow = Math.floor(totalAvailableWidth / blockWidth);
+    let maxBlockRows = Math.floor(totalAvailableHeight / blockHeight);
+    let maxBlockElements = maxBlocksPerRow * maxBlockRows;
+
+    while (maxBlockElements > blocks.length) {
+        maxBlockRows--;
+        maxBlockElements = maxBlocksPerRow * maxBlockRows;
+    }
+
+    return maxBlockElements;
+}
+
+function refreshBlocksDisplay() {
+    const blockElements = $('.block');
+    const currentlyDisplayedBlockCount = blockElements.length;
+    const maxBlockElements = calculateMaximumNumberOfBlocksToDisplay(blockElements);
+
+    if (currentlyDisplayedBlockCount > maxBlockElements) {
+        // remove the last x blocks
+        for (let i = currentlyDisplayedBlockCount; i > maxBlockElements; i--) {
+            blockElements[i - 1].remove();
+        }
+    }
+    else {
+        // add more blocks to fill display
+        for (let i = currentlyDisplayedBlockCount; i < maxBlockElements; i++) {
+            const newBlockElement = newBlockHtmlElement(blocks[i]);
+            blocksHolder.append(newBlockElement);
+        }
+    }
+
+    setupTooltips();
+}
+
+async function displayBlocks() {
```

Comment:
Dom manipulation is typically all synchronous, so there should be no need to make this an async function or to use set time outs.

Are you seeing any issues with just doing this?
```
function displayBlocks() {
    blocksHolder = $('.blocks-holder');
    blocksHolder.append(makeMempoolBlock(blocks[0]));
    const $mempoolBlock = $('.block');
    const maxBlockElements = calculateMaximumNumberOfBlocksToDisplay($mempoolBlock);
    const blocksHTML = _.times(maxBlockElements,(n) => {
        return newBlockHtmlElement(blocks[n+1])
    }).join("\n")
    blocksHolder.append(blocksHTML)
    setupTooltips();
    window.addEventListener('resize', refreshBlocksDisplay);
}
```

Reply:
Yeah, when I performed the block rendering synchronously, the page goes unresponsive for a brief moment. Sometimes it gets stuck completely and won't even allow me to refresh.

I use the sleep(ms) async function to pause temporarily after adding each block to the dom to give the browser some time to actually show the user something rather than keeping the page stale for a while

Reply:
Do you experience those same issues with the code I pasted?

Reply:
Performance is okay, but there's 1 issue. In order to calculate the number of blocks that can fit into available screen space, the mempool block is first added to the dom and its size (width and height) read.

The size (especially width) of the rectangular boxes is not fixed, only a minimum width is set. The browser does some calculation to determine what width to assign to each rectangular box such that no box is smaller than the minimum width and there's no unused white space after placing the boxes in each row.

The issue is, when we try to calculate the width of the mempool block shortly after appending it to the dom, the value read is incorrect. The width is roughly the same as the width of the entire blocks section.

Waiting a couple milliseconds gives the browser some time to properly size the mempool box and then the correct width can be read.